runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -970,9 +972,7 @@ public TimestampedValueWithId<T> next() {
             Iterables.mergeSorted(
                 ImmutableList.of(data, pendingInRange), TimestampedValueWithId.COMPARATOR);
         Iterable<TimestampedValue<T>> fullIterable =
-            Iterables.filter(
-                Iterables.transform(includingAdds, TimestampedValueWithId::getValue),
-                tv -> !pendingDeletes.contains(tv.getTimestamp()));
+            Iterables.transform(includingAdds, TimestampedValueWithId::getValue);
```

Comment:
What if a delete was issued after an add? It feels like the timing of overlapping deletes and adds also needs to be considered.

Reply:
This case was already correctly handled. When clearRange is called, any pending adds in that range are removed from pendingAdds:

https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java#L1001

Reply:
The last delete will win because of the filter clause at 950 and the last insert will win because of the transform that will merge the add at 975?